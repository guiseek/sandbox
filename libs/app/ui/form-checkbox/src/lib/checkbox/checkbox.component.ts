import { FormControlAccessor } from './../control-accessor'
import { takeUntil } from 'rxjs/operators'
import {
  AfterContentInit,
  AfterViewInit,
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  forwardRef,
  Injectable,
  Input,
  OnDestroy,
  OnInit,
  Optional,
  Output,
  Self,
  ViewChild,
} from '@angular/core'
import { AbstractControl, FormControl, NgControl, NG_VALUE_ACCESSOR } from '@angular/forms'
import { Subject } from 'rxjs'
import { EventInputTarget } from '../event-input'

@Injectable()
export class FormCheckbox extends FormControlAccessor {}

const FormCheckboxProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => FormCheckbox),
  multi: true,
}

let nextId = 0

@Component({
  selector: 'form-checkbox',
  template: `
    <div class="form-checkbox">
      <input
        [id]="id"
        type="checkbox"
        name="checkbox"
        [value]="value"
        [checked]="el?.checked"
        [attr.aria-checked]="el?.checked"
        [attr.disabled]="disabled"
        [formControl]="control"
        (change)="onChangeEvent($event)"
        #input
      />
      <label [for]="id">
        <ng-content></ng-content>

        <svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
          <path d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8" />
        </svg>
      </label>
    </div>
  `,
  styleUrls: ['../checkbox.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [FormCheckbox, FormCheckboxProvider],
})
export class CheckboxComponent extends FormCheckbox implements AfterContentInit, OnDestroy {
  destroy$ = new Subject<void>()

  @ViewChild('input', { static: true }) _el: ElementRef<HTMLInputElement>
  get el() {
    return this._el.nativeElement
  }

  private _id: string = `form-checkbox-${nextId++}`

  @Input()
  public set value(value: any) {
    this._value = value
  }
  public get value(): any {
    return this._value
  }

  @Input()
  public set id(value: string) {
    this._id = value
  }
  public get id(): string {
    return this._id
  }

  @Input()
  public set disabled(value: boolean) {
    this._disabled = value
  }

  @Output()
  valueChange = new EventEmitter<any>()

  @Output()
  checkedChange = new EventEmitter<boolean>()

  control: AbstractControl

  constructor(@Optional() @Self() public ngControl: NgControl) {
    super()
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control ? this.ngControl?.control : new FormControl()
  }

  onChangeEvent({ target }: EventInputTarget) {
    this.onChange(target.value)
    this.checkedChange.emit(target.checked)
    this.valueChange.emit(target.value)
  }

  ngOnDestroy() {
    this.destroy$.complete()
  }
}
