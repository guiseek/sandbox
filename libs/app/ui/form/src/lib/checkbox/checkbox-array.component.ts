import { AbstractControl, ControlContainer, FormArray, FormControl, NG_VALUE_ACCESSOR } from '@angular/forms'
import { CheckboxLabelComponent } from './checkbox-label.component'
import { CheckboxItemComponent } from './checkbox-item.component'
import { ControlAccessor } from '../control-accessor'
import { takeUntil } from 'rxjs/operators'
import { Subject } from 'rxjs'
import {
  Component,
  ChangeDetectionStrategy,
  Input,
  OnDestroy,
  ContentChildren,
  QueryList,
  AfterContentInit,
  Injectable,
  forwardRef,
  Output,
  EventEmitter,
  Optional,
  Self,
  AfterViewInit,
  ContentChild,
} from '@angular/core'

@Injectable()
export class CheckboxArrayAccessor extends ControlAccessor {}

const CheckboxArrayProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => CheckboxArrayComponent),
  multi: true,
}

let nextId = 0

@Component({
  selector: 'form-checkbox-array',
  template: `
    <div class="form-checkbox-array">
      <ng-content select="h3"></ng-content>
      <div role="group" [attr.aria-labelledby]="id">
        <ng-content></ng-content>
      </div>
    </div>
  `,
  providers: [CheckboxArrayAccessor, CheckboxArrayProvider],

  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CheckboxArrayComponent extends CheckboxArrayAccessor
  implements AfterContentInit, AfterViewInit, OnDestroy {
  destroy$ = new Subject<void>()

  @ContentChild(CheckboxLabelComponent) checkboxLabel: CheckboxLabelComponent

  @ContentChildren(CheckboxItemComponent) checkboxItems: QueryList<CheckboxItemComponent>

  private _id = `form-checkbox-array-${nextId++}`

  public get id(): string {
    return this._id
  }

  @Input() public set id(value: string) {
    this._id = value
  }

  @Output() valueChange = new EventEmitter<any>()

  control: AbstractControl

  get formArray() {
    return this.control as FormArray
  }

  constructor(@Optional() @Self() public ngControl: ControlContainer) {
    super()
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control ? this.ngControl?.control : new FormArray([])

    if (this.checkboxLabel) {
      this.checkboxLabel.el.id = this.id
    }

    if (this.checkboxItems?.length) {
      this.checkboxItems.map((item) => {
        item.checkboxChange
          .pipe(takeUntil(this.destroy$))

          .subscribe(this.onItemChange.bind(this))
      })
    }
  }

  onItemChange({ checked, value }) {
    if (checked) {
      this.formArray.push(new FormControl(value))
    } else {
      this.formArray.controls.forEach((ctrl, i) => {
        if (ctrl.value === value) {
          return this.formArray.removeAt(i)
        }
      })
    }
  }

  ngAfterViewInit() {
    this.control.valueChanges.pipe(takeUntil(this.destroy$)).subscribe((value) => this.onChange(value))
  }

  ngOnDestroy() {
    this.destroy$.complete()
  }
}
