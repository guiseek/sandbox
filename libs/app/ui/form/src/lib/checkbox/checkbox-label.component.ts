import { Component, ElementRef } from '@angular/core'

@Component({
  selector: `
    form-checkbox-label,
    h3[form-checkbox-label]
  `,
  template: `<ng-content></ng-content>`,
})
export class CheckboxLabelComponent {
  el: HTMLHeadingElement
  constructor(private _el: ElementRef<HTMLHeadingElement>) {
    this.el = this._el.nativeElement
  }
}
