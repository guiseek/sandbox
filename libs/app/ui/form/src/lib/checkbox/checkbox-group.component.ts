import { AbstractControl, ControlContainer, FormGroup, NG_VALUE_ACCESSOR } from '@angular/forms'
import { CheckboxLabelComponent } from './checkbox-label.component'
import { CheckboxComponent } from './checkbox.component'
import { ControlAccessor } from '../control-accessor'
import { takeUntil } from 'rxjs/operators'
import { Subject } from 'rxjs'
import {
  Component,
  ChangeDetectionStrategy,
  Input,
  OnDestroy,
  ContentChildren,
  QueryList,
  AfterContentInit,
  Injectable,
  forwardRef,
  Output,
  EventEmitter,
  Optional,
  Self,
  AfterViewInit,
  ContentChild,
} from '@angular/core'

@Injectable()
export class CheckboxGroupAccessor extends ControlAccessor {}

const CheckboxGroupProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => CheckboxGroupComponent),
  multi: true,
}

let nextId = 0

@Component({
  selector: 'form-checkbox-group',

  template: `
    <div class="form-checkbox-group">
      <ng-content select="h3"></ng-content>
      <div role="group" [attr.aria-labelledby]="id">
        <ng-content></ng-content>
      </div>
    </div>
  `,
  styles: [
    `
      :host.bordered .form-checkbox-group {
        border: 2px solid rgba(255, 255, 255, 0.1);
        border-radius: 12px;
        padding: 12px 12px 4px;
      }
    `,
  ],
  providers: [CheckboxGroupAccessor, CheckboxGroupProvider],

  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CheckboxGroupComponent extends CheckboxGroupAccessor
  implements AfterContentInit, AfterViewInit, OnDestroy {
  destroy$ = new Subject<void>()

  @ContentChild(CheckboxLabelComponent) checkboxLabel: CheckboxLabelComponent

  @ContentChildren(CheckboxComponent) checkboxes: QueryList<CheckboxComponent>

  private _id = `form-checkbox-group-${nextId++}`

  public get id(): string {
    return this._id
  }

  @Input() public set id(value: string) {
    this._id = value
  }

  @Output() valueChange = new EventEmitter<any>()

  control: AbstractControl

  get formGroup() {
    return this.control as FormGroup
  }

  constructor(@Optional() @Self() public ngControl: ControlContainer) {
    super()
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control ? this.ngControl?.control : new FormGroup({})

    if (this.checkboxLabel) {
      this.checkboxLabel.el.id = this.id
    }

    if (this.checkboxes?.length) {
      this.checkboxes.map((item) => {
        item.checkedChange
          .pipe(takeUntil(this.destroy$))

          .subscribe(this.onCheckboxChecked.bind(this))
      })
    }
  }

  onCheckboxChecked(item: CheckboxComponent) {
    if (item.el.value !== 'undefined' && item.el.checked) {
      item.control.patchValue(item.el.value)
    }
  }

  ngAfterViewInit() {
    this.control.valueChanges.pipe(takeUntil(this.destroy$)).subscribe((value) => this.onChange(value))
  }

  ngOnDestroy() {
    this.destroy$.complete()
  }
}
