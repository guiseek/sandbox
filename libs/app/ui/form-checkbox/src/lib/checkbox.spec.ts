import { Component } from '@angular/core'
import { ReactiveFormsModule, FormControl, FormArray } from '@angular/forms'
import { Spectator, SpectatorHost, createHostFactory, createComponentFactory } from '@ngneat/spectator'

import { CheckboxComponent } from './checkbox/checkbox.component'

const VIEWS = {
  onlyCheck: `<form-checkbox [formControl]="check"></form-checkbox>`,
  withLabel: `<form-checkbox [formControl]="check">Label</form-checkbox>`,
  withCustomId: `<form-checkbox id="x" [formControl]="check"></form-checkbox>`,
  groupItems: `
    <form-checkbox-group formArrayName="items">
      <form-checkbox-item value="value1">Valor 1</form-checkbox-item>
      <form-checkbox-item value="value2">Valor 2</form-checkbox-item>
      <form-checkbox-item value="value3">Valor 3</form-checkbox-item>
    </form-checkbox-group>
  `,
}

@Component({ selector: 'form-custom-form', template: '' })
class CustomFormComponent {
  check = new FormControl()
}

describe('CheckboxComponent', () => {
  describe('Host', () => {
    let spectator: SpectatorHost<CheckboxComponent, CustomFormComponent>
    const createHost = createHostFactory({
      component: CheckboxComponent,
      host: CustomFormComponent,
      imports: [ReactiveFormsModule],
    })

    describe('Views', () => {
      it('should display only input', () => {
        spectator = createHost(VIEWS.onlyCheck)
        expect(spectator.query('.form-checkbox')).toMatchSnapshot()
      })

      it('should display checkbox with label', () => {
        spectator = createHost(VIEWS.withLabel)
        expect(spectator.query('.form-checkbox')).toMatchSnapshot()
      })

      it('should display checkbox with custom id', () => {
        spectator = createHost(VIEWS.withCustomId)
        expect(spectator.query('.form-checkbox')).toMatchSnapshot()
      })
    })

    describe('Props', () => {
      it('should create an instance', () => {
        spectator = createHost(VIEWS.withCustomId)
        const input = spectator.query('.form-checkbox > input')
        expect(input.getAttribute('id')).toEqual('x')
      })
    })
  })

  describe('Forms', () => {
    let spectator: Spectator<CheckboxComponent>
    const createComponent = createComponentFactory({
      component: CheckboxComponent,
      imports: [ReactiveFormsModule],
    })

    it('should create', () => {
      spectator = createComponent()
      expect(spectator.component).toBeTruthy()
    })

    it('should value changed if set value', () => {
      spectator.component.control.setValue('abc')
      expect(spectator.component.control.value).toBe('abc')
    })

    it('should value changed if set value', async () => {
      spyOn(spectator.component.valueChange, 'emit')
      spectator.component.control.valueChanges.subscribe(() => {
        expect(spectator.component.onChangeEvent).toBeCalledTimes(1)
        expect(spectator.component.valueChange.emit).toBeCalledTimes(1)
        expect(spectator.component.checkedChange.emit).toBeCalledTimes(1)
      })
      spectator.component.control.setValue('qwe')
    })
  })
})
