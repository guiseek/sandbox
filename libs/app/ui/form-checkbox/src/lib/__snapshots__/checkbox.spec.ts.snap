// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxComponent Host Views should display checkbox with custom id 1`] = `
<div
  class="form-checkbox"
>
  <input
    aria-checked="false"
    class="ng-untouched ng-pristine ng-valid"
    id="x"
    name="checkbox"
    type="checkbox"
    value="undefined"
  />
  <label
    for="x"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`CheckboxComponent Host Views should display checkbox with label 1`] = `
<div
  class="form-checkbox"
>
  <input
    aria-checked="false"
    class="ng-untouched ng-pristine ng-valid"
    id="form-checkbox-1"
    name="checkbox"
    type="checkbox"
    value="undefined"
  />
  <label
    for="form-checkbox-1"
  >
    Label
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`CheckboxComponent Host Views should display only input 1`] = `
<div
  class="form-checkbox"
>
  <input
    aria-checked="false"
    class="ng-untouched ng-pristine ng-valid"
    id="form-checkbox-0"
    name="checkbox"
    type="checkbox"
    value="undefined"
  />
  <label
    for="form-checkbox-0"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;
